doc/api/async_hooks.md

```diff
@@ -804,6 +804,186 @@ for (let i = 0; i < 10; i++) {
 }
 ```
 
+## Class: AsyncLocalStorage
+<!-- YAML
+added: REPLACEME
+-->
+
+This class is used to create asynchronous state between callbacks and promise
+chains. It allows you to keep variables throughout the lifetime of a web request
+or any other asynchronous duration. It is similar to thread local storage
+ in other languages.
+
+The following example builds a logger that will always know the current HTTP
+request and use it to display enhanced logs without needing to explicitly pass
+the current HTTP request to it.
+
+```js
+const { AsyncLocalStorage } = require('async_hooks');
+const http = require('http');
+
+const kReq = 'CURRENT_REQUEST';
+const asyncLocalStorage = new AsyncLocalStorage();
+
+function log(...args) {
+  const store = asyncLocalStorage.getStore();
+  // Make sure the store exists and it contains a request.
+  if (store && store.has(kReq)) {
+    const req = store.get(kReq);
+    // Prints `GET /items ERR could not do something
+    console.log(req.method, req.url, ...args);
+  } else {
+    console.log(...args);
+  }
+}
+
+http.createServer((request, response) => {
+  asyncLocalStorage.run(() => {
+    const store = asyncLocalStorage.getStore();
+    store.set(kReq, request);
+    someAsyncOperation((err, result) => {
+      if (err) {
+        log('ERR', err.message);
+      }
+    });
+  });
+})
+.listen(8080);
+```
+
+When having multiple instance of `AsyncLocalStorage`, they are independant
+from each others. It is safe to instanciate this class multiple times.
+
+### Choosing between asynchronous and synchronous methods
+
+#### When to choose `run`
+This method is asynchronous. It is called with a callback function that
```

Comment:
The organization of the doc here is a bit confusing. Here it talks about when to start using `run()` but the method itself is not introduced until later in the doc. Specifically, in the statement `This method is asynchronous...`, it's not clear *what* method

Reply:
Agreed, I can move this part after the API documentaiton part.

Reply:
Done

Reply:
Done